ecs-agent/tmds/handlers/fault/v1/handlers/middlewares.go

```diff
@@ -0,0 +1,71 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package handlers
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/ecs-agent/logger"
+	"github.com/aws/amazon-ecs-agent/ecs-agent/metrics"
+)
+
+// responseWriterWrapper is a wrapper of http.ResponseWriter.
+type responseWriterWrapper struct {
+	http.ResponseWriter
+	status int
+}
+
+func (w *responseWriterWrapper) WriteHeader(status int) {
+	w.status = status
+	w.ResponseWriter.WriteHeader(status)
+}
+
+func TelemetryMiddleware(
+	next http.Handler,
+	mf metrics.EntryFactory,
+	faultOperation, faultType string) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		startTime := time.Now()
+		rw := &responseWriterWrapper{ResponseWriter: w}
+
+		// Call the next handler in the chain
+		next.ServeHTTP(rw, r)
+
+		durationInMs := time.Since(startTime).Milliseconds()
+		loggerFields := logger.Fields{
+			"StatusCode":   rw.status,
+			"DurationInMs": durationInMs,
+			"Request":      r.URL.Path,
+		}
+		logger.Info("Complete the request", loggerFields)
+		// Emit a duration metric.
+		mf.New(fmt.Sprintf("MetadataServer.%s%sDuration", faultOperation, faultType)).
+			WithFields(loggerFields).
+			WithGauge(durationInMs).
+			Done(nil)
+		// Emit a metric when the request fails.
+		if rw.status >= 500 && rw.status < 600 {
```

Comment:
Q: Can there ever be a case where the response/status code be > 600?

Reply:
It won't be. https://developer.mozilla.org/en-US/docs/Web/HTTP/Status

Reply:
right, was wondering whether `rw.status < 600` is needed here. this is a non-blocking nit

Reply:
I think that should be fine. Since we are tracking server error only which is between 500 - 600.